### PR TITLE
Create datasets with a function

### PR DESCRIPTION
**Pull request by ivanzvonkov**
- Removes `RawLabels` class in favor of creating a `ExampleDataset(LabeledDataset)` class which overrides `load_labels()`
- Adds `verify_df()` for checking the integrity of the dataframe

Prerequisite for easily sharing processed datasets across projects

### CODE REVIEW

File: .github/workflows/buildings-example-test.yaml [hunk 1]
```diff
@@ -29,9 +29,7 @@ jobs:
       env:
         # https://dvc.org/doc/user-guide/setup-google-drive-remote#authorization
         GDRIVE_CREDENTIALS_DATA: ${{ secrets.GDRIVE_CREDENTIALS_DATA }}
-      run: |
-        dvc pull $(openmapflow datapath DATASETS) -f
-        dvc pull $(openmapflow datapath MODELS) -f
+      run: dvc pull -f
```
Comment (ivanzvonkov) [inline .github/workflows/buildings-example-test.yaml:32]: Pulls in raw data to test the custom `load_labels()` functions

File: buildings-example/datasets.py [hunk 2]
```diff
@@ -1,44 +1,54 @@
...
+class GeowikiLandcover2017(LabeledDataset):
+    def load_labels(self):
+        df = pd.read_csv(raw_dir / "loc_all_2.txt", sep="\t")
+        df = df[(df.sumcrop / 100) > 0.5].copy()
+        df.rename(columns={"loc_cent_Y": LAT, "loc_cent_X": LON}, inplace=True)
+        df[START] = date(2017, 1, 1)
+        df[END] = date(2018, 12, 31)
+        df[label_col] = 0.0
+        df[SUBSET] = train_val_test_split(index=df.index, val=0.05, test=0.05)
+        return df
```
Comment (ivanzvonkov) [inline buildings-example/datasets.py:45]: Here's what the new datasets look like

File: buildings-example/datasets.py [hunk 3]
```diff
@@ -1,44 +1,54 @@
...
+
+from openmapflow.config import PROJECT_ROOT, DataPaths
+from openmapflow.constants import END, LAT, LON, START, SUBSET
+from openmapflow.label_utils import train_val_test_split
 from openmapflow.labeled_dataset import LabeledDataset, create_datasets
-from openmapflow.labeled_dataset_custom import CustomLabeledDataset
-from openmapflow.raw_labels import RawLabels
+
+raw_dir = PROJECT_ROOT / DataPaths.RAW_LABELS
+label_col = "is_building"
```
Comment (ivanzvonkov) [inline buildings-example/datasets.py:15]: A label column helps centralize which column will have the labels but it isn't absolutely necessary.

File: crop-mask-example/data/report.txt [hunk 4]
```diff
@@ -1,19 +1,17 @@
 DATASET REPORT (autogenerated, do not edit directly)
 
-geowiki_landcover_2017 (Timesteps: 24)
+GeowikiLandcover2017 (Timesteps: 24)
 ----------------------------------------------------------------------------
 eo_data_complete          34270
 eo_data_export_failed       841
 eo_data_missing_values      379
-eo_data_skipped             376
```
Comment (ivanzvonkov) [inline crop-mask-example/data/report.txt:8]: These are the 0.5 class probability values that are thrown out in `load_labels()` now

File: openmapflow/constants.py [hunk 5]
```diff
@@ -13,7 +13,7 @@
 TEMPLATE_README = TEMPLATES_DIR / "README.md"
 TEMPLATE_DEPLOY_YML = TEMPLATES_DIR / "github-deploy.yaml"
 TEMPLATE_TEST_YML = TEMPLATES_DIR / "github-test.yaml"
-VERSION = "0.1.6rc4"
+VERSION = "0.2.0rc1"
```
Comment (ivanzvonkov) [inline openmapflow/constants.py:16]: This is a breaking change but I am only updating the minor version since the package is still more or less under development

File: openmapflow/datasets.py [hunk 6]
```diff
@@ -1,41 +1,33 @@
...
+
+label_col = CLASS_PROB
+
+
+class GeowikiLandcover2017(LabeledDataset):
+    def load_labels(self):
+        # Source: https://github.com/nasaharvest/crop-mask/blob/master/datasets.py
+        df = pd.read_csv(f"{gcloud_url}/crop/geowiki_landcover_2017.csv")
+        df = df[df[label_col] != 0.5].copy()
+        return df
```
Comment (ivanzvonkov) [inline openmapflow/datasets.py:16]: If a dataset has already been processed somewhere else and uploaded to Google Cloud it can be loaded directly

File: openmapflow/ee_exporter.py [hunk 7]
```diff
@@ -156,6 +156,10 @@ def _export_for_polygon(
         ):
             return True
 
+        # Check if task is already started in EarthEngine
+        if not test and description in self.ee_task_list:
+            return True
```
Comment (ivanzvonkov) [inline openmapflow/ee_exporter.py:161]: Allow ee check for training data removed by #104

File: openmapflow/label_utils.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    def _get_points(polygon, samples: int):
+        x_min, y_min, x_max, y_max = polygon.bounds
+        x = np.random.uniform(x_min, x_max, samples)
+        y = np.random.uniform(y_min, y_max, samples)
+        gdf_points = gpd.GeoSeries(gpd.points_from_xy(x, y))
+        gdf_points = gdf_points[gdf_points.within(polygon)]
+        return gdf_points
+
+    list_of_points = np.vectorize(_get_points)(df.geometry, df.samples)
+    return gpd.GeoDataFrame(geometry=pd.concat(list_of_points, ignore_index=True))
```
Comment (ivanzvonkov) [inline openmapflow/label_utils.py:60]: Utility functions that can be used when loading labels

File: openmapflow/labeled_dataset.py [hunk 9]
```diff
@@ -60,16 +218,99 @@ def _label_eo_counts(df: pd.DataFrame) -> str:
...
-            text += (
-                f"\u2714 {subset} amount: {labels_in_subset}, "
-                + f"positive class: {positive_class_percentage:.1%}\n"
-            )
+            text += f"\u2714 {subset} amount: {labels_in_subset}"
+            if CLASS_PROB in df.columns:
+                positive_class_percentage = (
+                    df[df[SUBSET] == subset][CLASS_PROB] > 0.5
+                ).sum() / labels_in_subset
+                text += f", positive class: {positive_class_percentage:.1%}"
```
Comment (ivanzvonkov) [inline openmapflow/labeled_dataset.py:226]: Only add this text is CLASS_PROB column is available

File: openmapflow/labeled_dataset.py [hunk 10]
```diff
@@ -60,16 +218,99 @@ def _label_eo_counts(df: pd.DataFrame) -> str:
...
+    no_duplicates = check(
+        ~df.duplicated([LON, LAT, START, END]).any(), "No duplicates\n"
+    )
+    return (
+        lat_col_checks
+        and lon_col_checks
+        and subset_col_checks
+        and date_col_checks
+        and no_duplicates
+    )
```
Comment (ivanzvonkov) [inline openmapflow/labeled_dataset.py:311]: Checks that the df outputted by custom `load_labels()` is a good df

File: openmapflow/labeled_dataset.py [hunk 11]
```diff
@@ -79,53 +320,204 @@ class LabeledDataset:
...
         df = df[clean_df_condition(df)].copy()
         if check_eo_data and df[EO_DATA].isnull().any():
             raise ValueError(
-                f"{self.dataset} has missing earth observation data, "
+                f"{self.name} has missing earth observation data, "
                 + "run openmapflow create-datasets"
             )
-        if not skip_to_np:
-            df[EO_DATA] = df[EO_DATA].apply(_to_np)
+        if to_np:
```
Comment (ivanzvonkov) [inline openmapflow/labeled_dataset.py:372]: Changed default to not convert to_np

File: openmapflow/labeled_dataset.py [hunk 12]
```diff
@@ -79,53 +320,204 @@ class LabeledDataset:
...
+        for col in [EO_DATA, EO_LAT, EO_LON, EO_FILE]:
+            if col not in df.columns:
+                df[col] = None
+
+        df = df.round({LON: 8, LAT: 8})
+        df[START] = pd.to_datetime(df[START]).dt.strftime("%Y-%m-%d")
+        df[END] = pd.to_datetime(df[END]).dt.strftime("%Y-%m-%d")
+        return df
+
+    def create_dataset(self, force: bool = False):
```
Comment (ivanzvonkov) [inline openmapflow/labeled_dataset.py:393]: Moved from custom_labeled_dataset.py

File: openmapflow/pytorch_dataset.py [hunk 13]
```diff
@@ -1,247 +1,22 @@
...
-
-def _is_local(df: pd.DataFrame, bbox: Optional[BBox]) -> Union[bool, pd.Series]:
-    if bbox is None:
-        return True
-    return (
-        (df[LAT] >= bbox.min_lat)
-        & (df[LAT] <= bbox.max_lat)
-        & (df[LON] >= bbox.min_lon)
-        & (df[LON] <= bbox.max_lon)
-    )
```
Comment (ivanzvonkov) [inline openmapflow/pytorch_dataset.py:26]: All these functions are ones the user could implement if need be, they are not absolutely necessary for a MVP OpenMapFlow project

File: openmapflow/train_utils.py [hunk 14]
```diff
@@ -16,3 +18,46 @@ def generate_model_name(val_df: pd.DataFrame, start_month: str) -> str:
...
+def get_x_y(
+    df: pd.DataFrame, label_col: str, start_month: str, input_months: int
+) -> Tuple[List[np.ndarray], List[float]]:
+    i = MONTHS.index(start_month)
+
+    def to_numpy(x: str):
+        return str_to_np(x)[i : i + input_months, :]  # noqa
+
+    tqdm.pandas()
+    return df[EO_DATA].progress_apply(to_numpy).to_list(), df[label_col].to_list()
```
Comment (ivanzvonkov) [inline openmapflow/train_utils.py:63]: These utility functions take care of the work that was done inside the `PyTorchDataset` before

File: setup.py [hunk 15]
```diff
@@ -36,7 +36,13 @@
         ]
     },
     scripts=["openmapflow/scripts/openmapflow", "openmapflow/scripts/deploy.sh"],
-    install_requires=["numpy", "pandas==1.3.5", "pyyaml>=6.0", "requests", "tqdm"],
+    install_requires=[
+        "numpy",
+        "pandas==1.3.5",
+        "pyyaml>=6.0",
+        "requests",
+        "tqdm>=4.9.0",
```
Comment (ivanzvonkov) [inline setup.py:44]: This version is needed for `progress_apply()`

**Comment by ivanzvonkov** [comment]
Process:
1. Merge and publish a release candidate
2. Test and update notebooks 
3. Publish real release

File: README.md [hunk 16]
```diff
@@ -230,8 +234,8 @@ Only available through above Colab notebook. Cloud Architecture must be deployed
 
 # Accessing existing datasets
 ```python
-from openmapflow.datasets import togo_crop_2019
-df = togo_crop_2019.load_df()
+from openmapflow.datasets import TogoCrop2019
+df = TogoCrop2019().load_df(to_np=True)
 x = togo_crop_2019.iloc[0]["eo_data"]
```
Comment (gabrieltseng) [inline README.md:239]: Should this line be updated as well?

Comment (gabrieltseng) [reply README.md:239]: (and the line below)

File: openmapflow/datasets.py (hunk 6 above)
Comment (gabrieltseng) [reply openmapflow/datasets.py:16]: Would it make sense to migrate the processing code here anyway? and the raw labels into dvc?

Otherwise, if cropharvest migrates to using this dataset object the code used to go from `raw_data` -> this dataset will get lost.

File: openmapflow/labeled_dataset.py [hunk 17]
```diff
@@ -79,53 +320,204 @@ class LabeledDataset:
...
-    dataset: str = ""
-    country: str = ""
-    label_type: str = "binary"
-    license: str = ""
-    source: str = ""
-    dataset_dir: Path = PROJECT_ROOT / dp.DATASETS
-
     def __post_init__(self):
-        self.df_path = self.dataset_dir / (self.dataset + ".csv")
+        np.random.seed(SEED)
```
Comment (gabrieltseng) [inline openmapflow/labeled_dataset.py:326]: Do we want to be reseeding stuff every time a class is initialized? 

Or do we want to initialize stuff once on initial import? with some global `HAS_BEEN_SEEDED` variable

File: openmapflow/labeled_dataset.py (hunk 11 above)
Comment (gabrieltseng) [reply openmapflow/labeled_dataset.py:372]: why? Was it taking too long?

Seems more useful to have usable np arrays by default.

**Review by gabrieltseng — APPROVED** [review]
LGTM :) Just a few nits

File: openmapflow/labeled_dataset.py (hunk 11 above)
Comment (ivanzvonkov) [reply openmapflow/labeled_dataset.py:372]: These are the primary dataframe use cases I can think of:
1. Exploratory data analysis (one or a couple of rows at a time)
2. Training ( >80% of the rows at a time)
3. Evaluation (<20% of the rows at a time)

Loading all Geowiki points into memory takes 25s and it's only necessary for 2. 
It's easier to just load what's needed for 1 and 3, hence the default

image

File: README.md (hunk 16 above)
Comment (ivanzvonkov) [reply README.md]: ah thanks!

File: openmapflow/labeled_dataset.py (hunk 17 above)
Comment (ivanzvonkov) [reply openmapflow/labeled_dataset.py]: Moved to top of file

**Comment by ivanzvonkov** [comment]
image

(Can't reply to this directly on the UI so taking a screenshot)

I think datasets should generally be processed and stored where ever they are primarily used. For example, LEM datasets should be processed and stored in the LEM repo. If datasets have a dual purpose e.g. training new models for maps (e.g. timl) and being used inside benchmarks (e.g. cropharvest) then it makes sense to store the datasets close to the benchmark code. 
The datasets here are just used for examples which isn't the primary usage of the datasets so I don't think the processing code should go here. 

The idea of doing the URL read is that it makes it possible to process data in CropHarvest and use it in TIML without updating Zenodo or the Python package.

@gabrieltseng I may have missed your concern so let me know if that is the case

**Comment by gabrieltseng** [comment]
> @gabrieltseng I may have missed your concern so let me know if that is the case

@ivanzvonkov that makes sense !
